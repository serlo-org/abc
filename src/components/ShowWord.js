import { addIndex, map, toUpper } from 'ramda'
import React from 'react'
import { View, Text } from 'react-native'

import speakerImage from '../assets/images/speaker.png'

import { RoundImageWithButton } from '../components/Components'
import * as SoundUtils from '../utils/SoundUtils'

const mapIndexed = addIndex(map)

const hightlightStyle = {
  backgroundColor: 'rgba(255,255,255,0.42)',
  borderRadius: 20
}

const ShowWord = ({ image, sound, text, letter }) => {
  const short = SoundUtils.getSoundWithLength(sound, 'short')
  const long = SoundUtils.getSoundWithLength(sound, 'long')

  const play = () => {
    short.play((success) => {
      if (success) {
        setTimeout(() => long.play(), 1000)
      }
    })
  }

  const letters = mapIndexed(
    (char, key) => (
      <View
          key={key}
          style={[
            { padding: 5 },
            (toUpper(char) === toUpper(letter)) ? hightlightStyle : null
          ]}
      >
        <Text style={{ color: '#fff', fontSize: 40, fontWeight: 'bold' }}>{char}</Text>
      </View>
      ),
    text
  )

  return (
    <View style={{
      flex: 1,
      backgroundColor: '#00B4D5',
      alignItems: 'center',
      justifyContent: 'space-around'
    }}>
      <RoundImageWithButton
        image={image}
        imageSize={200}
        icon={speakerImage}
        buttonSize={40}
        onPress={play}
      />
      <View style={{ flexDirection: 'row' }}>
        {letters}
      </View>
    </View>
  )
}

export default ShowWord
