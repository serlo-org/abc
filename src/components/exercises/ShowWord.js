import { addIndex, forEach, map, toUpper } from 'ramda';
import React from 'react';
import { View, Text } from 'react-native';

import speakerImage from '../../assets/images/speaker.png';
import { loadSounds } from '../helpers/audio';
import { RoundImageWithButton } from '../Components';

const mapIndexed = addIndex(map);

const highlightStyle = {
  backgroundColor: 'rgba(255,255,255,0.42)',
  borderRadius: 20
};

const ShowWord = ({ image, sounds, text, letter }) => {
  const play = () => {
    const playAll = ([sound, ...rest]) => {
      sound.playAsync();
      sound.setPlaybackFinishedCallback(() => {
        sound.setPositionAsync(0);
        if (rest.length > 0) {
          setTimeout(() => playAll(rest), 1000);
        }
      });
    };
    playAll(sounds);
  };

  const letters = mapIndexed(
    (char, key) => (
      <View
        key={key}
        style={[
          { padding: 5 },
          toUpper(char) === toUpper(letter) ? highlightStyle : null
        ]}
      >
        <Text style={{ color: '#fff', fontSize: 40, fontWeight: 'bold' }}>
          {char}
        </Text>
      </View>
    ),
    text
  );

  return (
    <View
      style={{
        flex: 1,
        backgroundColor: '#00B4D5',
        alignItems: 'center',
        justifyContent: 'space-around'
      }}
    >
      <RoundImageWithButton
        image={image}
        imageSize={200}
        icon={speakerImage}
        buttonSize={40}
        onPress={play}
      />
      <View style={{ flexDirection: 'row' }}>
        {letters}
      </View>
    </View>
  );
};

export default loadSounds(ShowWord);
