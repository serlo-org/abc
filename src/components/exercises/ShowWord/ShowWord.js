import { addIndex, map, toUpper } from 'ramda';
import React from 'react';
import { View, Text } from 'react-native';

import repeatIcon from '../../../assets/images/repeat.png';
import { getWord } from '../../../helpers/words';
import { WHITE_TRANSPARENT } from '../../../styles/colors';
import { DEFAULT } from '../../../styles/text';
import IconWithBackground from '../../common/IconWithBackground';
import WordImageWithSounds from '../../common/WordImageWithSounds';

const mapIndexed = addIndex(map);

const highlightStyle = {
  backgroundColor: WHITE_TRANSPARENT,
  borderRadius: 20
};

const ShowWord = ({ word, letter, sound, setState, isRepeat }) => {
  const letters = mapIndexed(
    (char, key) => (
      <View
        key={key}
        style={[
          { padding: 5 },
          toUpper(char) === toUpper(letter) ? highlightStyle : null
        ]}
      >
        <Text style={DEFAULT}>{char}</Text>
      </View>
    ),
    getWord(word)
  );

  return (
    <View
      style={{
        flex: 1,
        alignItems: 'center',
        justifyContent: 'space-around'
      }}
    >
      <WordImageWithSounds
        longSound
        isRepeat={isRepeat}
        word={word}
        onPlayStart={() => {
          if (isRepeat) {
            this.icon.unfocus();
            setState(false);
          }
        }}
        onPlayEnd={() => {
          if (isRepeat) {
            this.icon.focus();
            setState(true);
          }
        }}
      />
      <View style={{ flexDirection: 'row' }}>{letters}</View>
      <View height={80}>
        {isRepeat && (
          <IconWithBackground
            ref={view => {
              this.icon = view;
            }}
            icon={repeatIcon}
            size={40}
          />
        )}
      </View>
    </View>
  );
};

export default ShowWord;
