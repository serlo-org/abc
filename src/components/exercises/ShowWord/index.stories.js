import { storiesOf } from '@storybook/react-native';

import * as ShowWord from '.';
import { createStoriesFromFixtures } from '../helpers';

const story = storiesOf('exercises/ShowWord', module);

createStoriesFromFixtures(story, ShowWord);
