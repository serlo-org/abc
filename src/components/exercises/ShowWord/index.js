import { InfoScreen } from '../../../exercises';
import { getWordObject } from '../../../helpers/words';
import Component from './ShowWord';

export const Exercise = InfoScreen.Exercise;
export const fixtures = InfoScreen.createFixtures([
  {
    name: 'Ananas',
    props: {
      word: getWordObject('ananas'),
      letter: 'A'
    }
  }
]);
export { Component };
