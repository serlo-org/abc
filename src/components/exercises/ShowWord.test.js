import React from 'react';

import ShowWord from './ShowWord';

import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const tree = renderer.create(
    <ShowWord
      image={require('../../assets/images/ananas.jpg')}
      sounds={[
        require('../../assets/sounds/ananas_long.mp3'),
        require('../../assets/sounds/ananas_short.mp3')
      ]}
      text="Ananas"
      letter="A"
    />
  );
  expect(tree).toMatchSnapshot();
});
