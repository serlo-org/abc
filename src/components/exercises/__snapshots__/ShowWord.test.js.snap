// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00B4D5",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.05)",
            "borderRadius": 9999999,
            "margin": 20,
          },
          undefined,
        ]
      }
    >
      <Image
        resizeMode="cover"
        source={1}
        style={
          Object {
            "borderRadius": 100,
            "height": 200,
            "margin": 20,
            "width": 200,
          }
        }
      />
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": -70,
          "marginRight": 20,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "backgroundColor": "#73DBFF",
            "borderRadius": 9999,
            "elevation": 10,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 40,
              "resizeMode": "contain",
              "width": 40,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "rgba(255,255,255,0.42)",
            "borderRadius": 20,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        A
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        n
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "rgba(255,255,255,0.42)",
            "borderRadius": 20,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        a
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        n
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "rgba(255,255,255,0.42)",
            "borderRadius": 20,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        a
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          null,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 40,
            "fontWeight": "bold",
          }
        }
      >
        s
      </Text>
    </View>
  </View>
</View>
`;
